src/hotspot/share/opto/mempointer.hpp

```diff
@@ -0,0 +1,591 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_OPTO_MEMPOINTER_HPP
+#define SHARE_OPTO_MEMPOINTER_HPP
+
+#include "opto/memnode.hpp"
+#include "opto/noOverflowInt.hpp"
+
+// The MemPointer is a shared facility to parse pointers and check the aliasing of pointers,
+// e.g. checking if two stores are adjacent.
+//
+// -----------------------------------------------------------------------------------------
+//
+// Intuition and Examples:
+//   We parse / decompose pointers into a linear form:
+//
+//     pointer = con + sum_i(scale_i * variable_i)
+//
+//   The con and scale_i are compile-time constants (NoOverflowInt), and the variable_i are
+//   compile-time variables (C2 nodes).
+//
+//   For the MemPointer, we do not explicitly track base address. For Java heap pointers, the
+//   base address is just a variable. For native memory (C heap) pointers, the base address is
```

Comment:
May be: "For Java heap pointers, the  base address is just a variable  in summand with `scale` == 1.